### PR TITLE
Fix analyzation of individual paths

### PR DESCRIPTION
**Pull request by BlakeWilliams**
This re-introduces `requested_paths` to several files that gets passed
into `include_paths_builder`. This was removed in an earlier commit but
is necessary for specifying individual files or paths to analyze instead
of analyzing the entire directory.

### CODE REVIEW

File: lib/cc/analyzer/path_entries.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        end
+      end
+
+      private
+
+      def all_entries
+        Dir.glob("#{initial_path}/**/*", File::FNM_DOTMATCH).push(initial_path)
+      end
+
+      attr_reader :initial_path
```
Comment (pbrisbin) [inline lib/cc/analyzer/path_entries.rb]: :lipstick: order macros like this first after the `private` pragma

File: lib/cc/analyzer/path_patterns.rb [hunk 2]
```diff
@@ -10,30 +10,40 @@ def expanded
         @expanded ||= expand
       end
 
+      def match?(path)
+        @patterns.any? do |pattern|
+          value = glob_value(pattern)
+          File.fnmatch(pattern, path)
```
Comment (pbrisbin) [inline lib/cc/analyzer/path_patterns.rb]: Hi :) I assume you'll be ditching this in favor of `#expand` or `PathPatterns`?

Comment (BlakeWilliams) [inline lib/cc/analyzer/path_patterns.rb]: Welcome back! I do!

File: lib/cc/analyzer/path_filter.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        patterns = PathPatterns.new(globs)
+        @paths = paths.reject { |path| patterns.match?(pathpatterns.match?(path)) }
+        self
+      end
+
+      private
+
+      def unreadable_path_entries
+        @_unreadable_path_entries ||=
+          unreadable_paths.map { |path| PathEntries.new(path).entries }.flatten
```
Comment (pbrisbin) [inline lib/cc/analyzer/path_filter.rb]: `map.flatten` -> `flat_map`?

**Comment by pbrisbin** [comment]
CC issues look legit

File: lib/cc/analyzer/include_paths_builder.rb [hunk 4]
```diff
@@ -14,72 +15,36 @@ def initialize(cc_exclude_paths, cc_include_paths = [])
...
 
       def include_paths
         if @cc_include_paths.empty?
-          Dir.glob("**/*", File::FNM_DOTMATCH)
+          all_paths
         else
           @cc_include_paths.map do |path|
-            glob_or_include_path(path)
+            PathEntries.new(path).entries
           end.flatten
```
Comment (pbrisbin) [inline lib/cc/analyzer/include_paths_builder.rb]: `map.flatten` => `flat_map` again please.

File: lib/cc/analyzer/include_paths_builder.rb [hunk 5]
```diff
@@ -14,72 +15,36 @@ def initialize(cc_exclude_paths, cc_include_paths = [])
...
-
-      def raise_if_unreadable(path)
-        if !File.directory?(path) && !FileUtils.readable_by_all?(path)
-          raise CC::Analyzer::UnreadableFileError, "Can't read #{path}"
-        end
+      def all_paths
+        Dir.glob("*", File::FNM_DOTMATCH).
+          reject { |path| IncludePathsBuilder::IGNORE_PATHS.include?(path) }.
+          map { |path| PathEntries.new(path).entries }.
+          flatten
```
Comment (pbrisbin) [inline lib/cc/analyzer/include_paths_builder.rb]: `flat_map` again

File: lib/cc/analyzer/path_patterns.rb [hunk 6]
```diff
@@ -15,25 +19,28 @@ def expanded
...
 
         results.sort.uniq
       end
+
+      def glob_value(pattern)
+        # FIXME: there exists a temporary workaround whereby **-style globs
+        # are translated to **/*-style globs within cc-yaml's custom
+        # Glob#value method. It was thought that that would work correctly
+        # with Dir.glob but it turns out we have to actually invoke #value
+        # directory for this to work. We need to guard this on class (not
```
Comment (pbrisbin) [inline lib/cc/analyzer/path_patterns.rb]: This was my typo, but "directory" -> "directly"

**Comment by pbrisbin** [comment]
Love the various extracted classes. LGTM.

**Comment by ABaldwinHunter** [comment]
@blake looking now. Do `exclude_paths` still get respected when a path is passed to `codeclimate analyze <path1> <path2>`?

**Comment by BlakeWilliams** [comment]
Yeah, they should since it's only created with  `IncludePathsBuilder.new(exclude_paths, Array(@requested_paths)).build`

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams wait

**Comment by ABaldwinHunter** [comment]
what manual testing have you done? I tried it out just now and am afraid that I see duplicate paths behavior

**Comment by BlakeWilliams** [comment]
This doesn't fix duplicate paths. I'm not sure how to reproduce that one.

**Comment by ABaldwinHunter** [comment]
```
Ashleys-MacBook-Pro-2:railsgoat ashleybaldwin-hunter$ codeclimate analyze -e rubocop lib/
Starting analysis
Running rubocop: Done!

== lib/tasks/server.rake (6 issues) ==
6: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
6: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
6: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
16: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
16: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
16: `File.exists?` is deprecated in favor of `File.exist?`. [rubocop]
```

**Comment by BlakeWilliams** [comment]
I was thinking we could either merge this fix and start a new PR or figure out how to repro the duplication issue.

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams If we can't fix that problem, we should revert to before the minimization change

**Comment by BlakeWilliams** [comment]
I haven't looked into it yet since they're two different issues and this is the fix individual paths.

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams I think they're related - it's how do you pass `paths` to `codeclimate analyze` - whether they're files or directories

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams nice work.

In general, I feel like the original minimize paths builder feature rebuilt some functionality that already existed, and I question that decision but it feels like a separate issue. 

I really like the added tests here and I'm happy that this update fixes the broken behavior on master!

Can you compact the commits? Besides that LGTM

**Comment by ABaldwinHunter** [comment]
:shipit:

**Comment by BlakeWilliams** [comment]
It definitely rebuilds some functionality unintentionally, but we can revisit that later like you said.

I'll squash and merge. Thanks for the review!

**Pull request merged**